src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp

```diff
@@ -369,6 +380,242 @@ void CascadiaSettings::_LoadDynamicProfiles()
     }
 }
 
+// Method Description:
+// - Searches the local app data folder, global app data folder and app
+//   extensions for json stubs we should use to create new profiles,
+//   modify existing profiles or add new colour schemes
+// - If the user settings has any namespaces in the "disabledProfileSources"
+//   property, we'll ensure that the corresponding folders do not get searched
+void CascadiaSettings::_LoadProtoExtensions()
+{
+    // First, accumulate the namespaces the user wants to ignore
+    std::unordered_set<std::wstring> ignoredNamespaces;
+    const auto disabledProfileSources = CascadiaSettings::_GetDisabledProfileSourcesJsonObject(_userSettings);
+    if (disabledProfileSources.isArray())
+    {
+        for (const auto& json : disabledProfileSources)
+        {
+            ignoredNamespaces.emplace(JsonUtils::GetValue<std::wstring>(json));
+        }
+    }
+
+    // Search through the local app data folder
+    if (std::filesystem::exists(wil::ExpandEnvironmentStringsW<std::wstring>(LocalAppDataFolder.data())))
+    {
+        _ApplyJsonStubsHelper(LocalAppDataFolder, ignoredNamespaces);
+    }
+
+    // Search through the global app data folder
+    if (std::filesystem::exists(wil::ExpandEnvironmentStringsW<std::wstring>(ProgramDataFolder.data())))
+    {
+        _ApplyJsonStubsHelper(ProgramDataFolder, ignoredNamespaces);
+    }
+
+    // Search through app extensions
+    // Gets the catalog of extensions with the name "com.microsoft.windows.terminal.settings"
+    const auto catalog = Windows::ApplicationModel::AppExtensions::AppExtensionCatalog::Open(winrt::to_hstring(AppExtensionHostName));
+
+    // Extracting the list of extensions from the catalog is an async operation, but since
+    // we are on a UI thread - and so cannot call a blocking operation like .get() - we use
+    // a mutex and condition variable
+    std::condition_variable cv;
+    std::mutex mtx;
+    Windows::Foundation::Collections::IVectorView<Windows::ApplicationModel::AppExtensions::AppExtension> extensions;
+
+    auto lambda = [&]() -> winrt::fire_and_forget {
+        co_await resume_background();
+        const auto localExtensions = catalog.FindAllAsync().get();
+
+        std::unique_lock<std::mutex> lock{ mtx };
+        extensions = localExtensions;
+        cv.notify_all();
+    };
+
+    std::unique_lock<std::mutex> lock{ mtx };
+    lambda();
+    cv.wait(lock);

```

Comment:
I am 90% sure that there is a `wil::wait_for_completion` function that does effectively this in a one liner.

Reply:
@DHowett and I mused about this for half an hour and everything else is so much worse. Forget it.